### PR TITLE
Add conversion for DivOp ONNX to TOSA

### PR DESCRIPTION
**Pull request by sagarkulkarni19**

### CODE REVIEW

**Comment by sagarkulkarni19** [comment]
Hi @maxbartel , @tungld . Pinging for reviews. Thanks!

**Comment by AlexandreEichenberger** [comment]
updating to the latest, will run all CIs.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, assume someone familiar with TOSA can vouch for the logic of the sequences being generated.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 1]
```diff
@@ -82,4 +82,47 @@ func.func @test_sub_broadcast(%arg0: tensor<13x21x1xf32>, %arg1: tensor<1xf32>)
...
+}
+
+// -----
+
+func.func @test_div_broadcast(%arg0: tensor<13x21x1xi32>, %arg1: tensor<1xi32>) -> tensor<13x21x1xi32> {
+  %0 = "onnx.Div"(%arg0, %arg1) : (tensor<13x21x1xi32>, tensor<1xi32>) -> tensor<13x21x1xi32>
+  "func.return"(%0) : (tensor<13x21x1xi32>) -> ()
+// CHECK-LABEL:  func @test_div_broadcast
+// CHECK-SAME:   ([[PARAM_0_:%.+]]: tensor<13x21x1xi32>, [[PARAM_1_:%.+]]: tensor<1xi32>) -> tensor<13x21x1xi32> {
+// CHECK-NEXT:      [[VAR_0_:%.+]] = "tosa.div"([[PARAM_0_]], [[PARAM_1_]]) : (tensor<13x21x1xi32>, tensor<1xi32>) -> tensor<13x21x1xi32>
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:105]: Same here.

**Comment by AlexandreEichenberger** [comment]
If broadcasting is not supported, probably an error should be reported if the pattern is encountered.

`ONNXBroadcastOpShapeHelper` has a method `hasNoBroadcast(...)`, FYI.

**Comment by sagarkulkarni19** [comment]
Apologies for the delay. Will address these issues today.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 2]
```diff
@@ -82,4 +82,47 @@ func.func @test_sub_broadcast(%arg0: tensor<13x21x1xf32>, %arg1: tensor<1xf32>)
...
+
+// -----
+
+func.func @test_div_decomposed_broadcast(%arg0: tensor<13x21x1xf32>, %arg1: tensor<1xf32>) -> tensor<13x21x1xf32> {
+  %0 = "onnx.Div"(%arg0, %arg1) : (tensor<13x21x1xf32>, tensor<1xf32>) -> tensor<13x21x1xf32>
+  "func.return"(%0) : (tensor<13x21x1xf32>) -> ()
+// CHECK-LABEL:  func @test_div_decomposed_broadcast
+// CHECK-SAME:   ([[PARAM_0_:%.+]]: tensor<13x21x1xf32>, [[PARAM_1_:%.+]]: tensor<1xf32>) -> tensor<13x21x1xf32> {
+// CHECK-NEXT:      [[VAR_0_:%.+]] = "tosa.reciprocal"([[PARAM_1_]]) : (tensor<1xf32>) -> tensor<1xf32>
+// CHECK-NEXT:      [[VAR_1_:%.+]] = "tosa.mul"([[PARAM_0_]], [[VAR_0_]]) {shift = 0 : i32} : (tensor<13x21x1xf32>, tensor<1xf32>) -> tensor<13x21x1xf32>
```
[the remark this answers is not in the document]
Comment (sagarkulkarni19) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:127]: I see. Okay will do.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 1 above)
Comment (sagarkulkarni19) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:105]: Got it.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 2 above)
[the remark this answers is not in the document]
Comment (sagarkulkarni19) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:127]: Actually I have a questions about rank broadcasting after looking at other element wise operators: Why do operators like Add and Sub do not need a ReshapeOp?

[the remark this answers is not in the document]
Comment (maxbartel) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:127]: That is true. My new PR https://github.com/onnx/onnx-mlir/pull/2269 is going to fix this, hopefully I found all places where this was wrong.

[the remark this answers is not in the document]
Comment (sagarkulkarni19) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:127]: Sounds good. I'll wait for that PR to land and make changes to my MR accordingly.

**Comment by AlexandreEichenberger** [comment]
@sagarkulkarni19 any progress on this PR?

**Comment by sagarkulkarni19** [comment]
> @sagarkulkarni19 any progress on this PR?

Updating it today. Thanks for the ping, I missed max's reply.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 2 above)
[the remark this answers is not in the document]
Comment (sagarkulkarni19) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir]: Updated.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir (hunk 1 above)
Comment (sagarkulkarni19) [reply test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir]: Updated.

**Comment by sagarkulkarni19** [comment]
Rebased and addressed comments.

**Comment by AlexandreEichenberger** [comment]
rebased, will launch the CIs automatically

**Review by maxbartel — APPROVED** [review]
LGTM
@eric-k256 Wouldn't it be better to rename div to intdiv to match the specification? It would also make it clearer that it really only supports integers

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 3]
```diff
@@ -130,6 +130,33 @@ class ONNXReluOpLoweringToTOSA : public OpConversionPattern<ONNXReluOp> {
...
+      Value divOp = tosaBuilder.div(lhs, rhs);
+      rewriter.replaceOp(op, {divOp});
+    } else {
+      // If it is not a 32-bit signless integer, decompose ONNXDivOp into
+      // tosa::ReciprocalOp and tosa::MulOp
+      Value reciprocalOp = tosaBuilder.reciprocal(rhs);
+      Value mulOp = tosaBuilder.mul(lhs, reciprocalOp);
+      rewriter.replaceOp(op, {mulOp});
+    }
+    return success();
```
[the remark this answers is not in the document]
Comment (sagarkulkarni19) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp]: Done. Thanks for the suggestion.

**Comment by eric-k256** [comment]
> LGTM @eric-k256 Wouldn't it be better to rename `div` to `intdiv` to match the specification? It would also make it clearer that it really only supports integers

Yes, good point, I agree that the op should be renamed. I'll try to get that done soon.

**Comment by sagarkulkarni19** [comment]
Thanks for the review @maxbartel and @AlexandreEichenberger . Btw, would you prefer that I change the TosaBuilder function name from `div` to `intdiv` or is the current state good to merge?

File: src/Conversion/ONNXToTOSA/DialectBuilder.cpp [hunk 4]
```diff
@@ -182,6 +182,36 @@ Value TosaBuilder::mul(mlir::Value &lhs, mlir::Value &rhs, int32_t shift) {
       rewriter(), loc(), newValueType, lhs, rhs, shift);
 }
 
+Value TosaBuilder::div(mlir::Value &lhs, mlir::Value &rhs) {
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToTOSA/DialectBuilder.cpp:185]: There was the suggestion to rename `div` to `intdiv` or something of the sort; was the comment acted upon (saw a thumbs up, which seems to indicate that this was agreed? )

Comment (eric-k256) [reply src/Conversion/ONNXToTOSA/DialectBuilder.cpp:185]: My comment was about renaming the TOSA dialect op from DivOp to IntDivOp, but that needs to happen upstream and then updated in onnx-mlir along with the LLVM merge that includes the upstream change. We could change the method here to TosaBuilder::intdiv now to minimize the future work and be more clear at the TosaBuilder interface today.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

**Comment by sagarkulkarni19** [comment]
@AlexandreEichenberger @maxbartel should I push another rebase? If there are no other issues, maybe we can merge this?

**Comment by AlexandreEichenberger** [comment]
yes, rebasing now, can merge when it succeeds, if that is what you would like me to do

**Comment by sagarkulkarni19** [comment]
> yes, rebasing now, can merge when it succeeds, if that is what you would like me to do

Thanks @AlexandreEichenberger . Was just pinging to check if there are any remaining issues to address :-) I'm good with merging it if the pipeline succeeds. Thanks for the review!

**Pull request merged**